src/python/pants/util/meta.py

```diff
@@ -138,3 +139,41 @@ def new_setattr(self, key: str, value: Any) -> None:
   cls.__init__ = new_init  # type: ignore
   cls.__setattr__ = new_setattr  # type: ignore
   return cls
+
+
+class EnumMatchError(ValueError):
+  """Issue when using match() on an enum."""
+
+
+class InexhaustiveMatchError(EnumMatchError):
+  """Not all values of the enum specified in the pattern match."""
+
+
+class UnrecognizedMatchError(EnumMatchError):
+  """A value is used that is not a part of the enum."""
+
+
+_E = TypeVar('_E', bound='Enum')
+_V = TypeVar('_V')
+
+
+def match(enum_instance: Enum, enum_values_to_results: Mapping[_E, _V]) -> _V:
```

Comment:
Now that this is a top level function, I think we should rename this to `enum_match`. It makes it more clear what this refers to.

Reply:
Agreed. Or move this to pants.util.enums so that users can say:
```python
from pants.util import enums
...
enums.match(...)
```

It seems to me `meta` is a strange home. This is not really metaprogramming related at all, its just attempting to add a language feature / support a programming style python doesn't out of the box.

Reply:
I hadn't previously considered a distinction between metaprogramming and supporting a new language feature.